File: sourceCode_PhiSpaceST/scripts/00_setup/paths_loader.R
# PhiSpace ST path loader (R).
#
# Source this from any analysis script:
#   source("scripts/00_setup/paths_loader.R")
# It walks up from the current working directory to find
# `config/paths_template.yml`, reads `paths_local.yml` if present (else the
# template, with a warning), and exposes a `paths` named list.
#
# Convenience derivations:
#   paths$data_root   = file.path(paths$phispace_data_root, "data")
#   paths$output_root = file.path(paths$phispace_data_root, "output")

local({
  find_proj_root <- function(start = getwd()) {
    d <- normalizePath(start, mustWork = TRUE)
    repeat {
      if (file.exists(file.path(d, "config", "paths_template.yml"))) return(d)
      parent <- dirname(d)
      if (parent == d) {
        stop("paths_loader: could not find sourceCode_PhiSpaceST/config/paths_template.yml ",
             "by walking up from '", start, "'. Set the working directory to the ",
             "sourceCode_PhiSpaceST/ root (or anywhere inside it) before sourcing this file.")
      }
      d <- parent
    }
  }

  if (!requireNamespace("yaml", quietly = TRUE)) {
    stop("paths_loader: the 'yaml' package is required. Install with install.packages('yaml').")
  }

  proj <- find_proj_root()
  local_yml    <- file.path(proj, "config", "paths_local.yml")
  template_yml <- file.path(proj, "config", "paths_template.yml")
  cfg_path <- if (file.exists(local_yml)) local_yml else template_yml
  if (cfg_path == template_yml) {
    message("[paths_loader] Using paths_template.yml (placeholders). ",
            "Copy it to paths_local.yml and edit before running real analyses.")
  }

  cfg <- yaml::read_yaml(cfg_path)
  cfg$project_root <- proj
  if (!is.null(cfg$phispace_data_root)) {
    if (is.null(cfg$data_root))   cfg$data_root   <- file.path(cfg$phispace_data_root, "data")
    if (is.null(cfg$output_root)) cfg$output_root <- file.path(cfg$phispace_data_root, "output")
  }
  assign("paths", cfg, envir = parent.frame(2))
})

File: sourceCode_PhiSpaceST/scripts/00_setup/paths_loader.py
"""PhiSpace ST path loader (Python).

Usage from any analysis script::

    import sys
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "00_setup"))
    from paths_loader import paths

It walks up from the current working directory to find
``config/paths_template.yml``, reads ``paths_local.yml`` if present (else the
template, with a warning), and exposes a ``paths`` dict.

Convenience derivations::

    paths["data_root"]   = paths["phispace_data_root"] + "/data"
    paths["output_root"] = paths["phispace_data_root"] + "/output"
"""

from __future__ import annotations

import os
import sys
import warnings
from pathlib import Path

try:
    import yaml
except ImportError as exc:
    raise ImportError(
        "paths_loader: the 'pyyaml' package is required. "
        "Install with `pip install pyyaml`."
    ) from exc


def _find_proj_root(start: str | os.PathLike | None = None) -> Path:
    p = Path(start or os.getcwd()).resolve()
    while not (p / "config" / "paths_template.yml").exists():
        if p.parent == p:
            raise FileNotFoundError(
                "paths_loader: could not find sourceCode_PhiSpaceST/config/"
                "paths_template.yml by walking up from '"
                f"{Path(start or os.getcwd()).resolve()}'."
            )
        p = p.parent
    return p


def _load() -> dict:
    proj = _find_proj_root()
    local_yml = proj / "config" / "paths_local.yml"
    template_yml = proj / "config" / "paths_template.yml"
    cfg_path = local_yml if local_yml.exists() else template_yml
    if cfg_path == template_yml:
        warnings.warn(
            "[paths_loader] Using paths_template.yml (placeholders). "
            "Copy it to paths_local.yml and edit before running real analyses.",
            stacklevel=2,
        )
    with open(cfg_path) as f:
        cfg = yaml.safe_load(f) or {}
    cfg["project_root"] = str(proj)
    pdr = cfg.get("phispace_data_root")
    if pdr:
        cfg.setdefault("data_root", str(Path(pdr) / "data"))
        cfg.setdefault("output_root", str(Path(pdr) / "output"))
    return cfg


paths = _load()
